arduino-core/src/cc/arduino/packages/uploaders/SerialUploader.java

```diff
@@ -393,13 +374,27 @@ public boolean burnBootloader() throws Exception {
 
     new LoadVIDPIDSpecificPreferences().load(prefs);
 
-    String pattern = prefs.getOrExcept("erase.pattern");
-    String[] cmd = StringReplacer.formatAndSplit(pattern, prefs);
-    if (!executeUploadCommand(cmd))
+    if (!runCommand("erase.pattern", prefs))
       return false;
 
-    pattern = prefs.getOrExcept("bootloader.pattern");
-    cmd = StringReplacer.formatAndSplit(pattern, prefs);
-    return executeUploadCommand(cmd);
+    return runCommand("bootloader.pattern", prefs);
+  }
+
+  private boolean runCommand(String patternKey, PreferencesMap prefs) throws Exception, RunnerException {
+    try {
+      String pattern = prefs.getOrExcept(patternKey);
+      StringReplacer.checkIfRequiredKeyIsMissingOrExcept("serial.port", pattern, prefs);
```

Comment:
Is this the only line that can throw a PreferencesMapException? If so, wouldn't it be better to have a smaller try block for that (or perhaps even let checkIfRequiredKeyIsMissingOrExcept return a boolean, if all it does is except or not except)?

Reading this, I wonder: Shouldn't `StringReplacer.formatAndSplit` (etc.) just throw when they find any unset variable? Or does that break for optional variables in some cases?